### PR TITLE
Flame v1.2.0 migration

### PR DESCRIPTION
**Pull request by rozpo**
# Major description of changes:

## SpacescapeGame:
- Dead code **prepare**() and **onResize**() methods due fact that overrides no longer exists in v1.2.0

## Migrate JoystickComponent:
- joystickChangeDirectional() **removed**
- joystickAction() updated, now called by **ButtonComponent**
- Player.update() adapt new **position** calculation.
- Now joystick component is required for **Player** class.
- **ButtonComponent** added instead actions in JoystickComponent.

## KnowsGameSize:
- No longer used in code, remains as is as reference for the YouTube tutorial with note about it in code.
- Migrate for use **gameRef.size** instead.

## Player:
- **onLoad** method added.
- Fill **_playerData** with defaultData to make sure a null exception is not thrown due update in game.dart (**_playerScore.text**).

### CODE REVIEW

File: lib/game/bullet.dart [hunk 1]
```diff
@@ -28,17 +28,22 @@ class Bullet extends SpriteComponent with Hitbox, Collidable {
 
     // Adding a circular hitbox with radius as 0.4 times
     //  the smallest dimension of this components size.
-    final shape = HitboxCircle(definition: 0.4);
-    addShape(shape);
+    final shape = CircleHitbox.relative(
+      0.4,
+      parentSize: this.size,
+      position: Vector2(size.x / 2, size.y / 2),
```
Comment (ufrshubham) [inline lib/game/bullet.dart]: Can be written as `position: size/2` for simplicity.

File: lib/game/enemy.dart [hunk 2]
```diff
@@ -95,8 +94,13 @@ class Enemy extends SpriteComponent
 
     // Adding a circular hitbox with radius as 0.8 times
     // the smallest dimension of this components size.
-    final shape = HitboxCircle(definition: 0.8);
-    addShape(shape);
+    final shape = CircleHitbox.relative(
+      0.8,
+      parentSize: this.size,
+      position: Vector2(size.x / 2, size.y / 2),
```
Comment (ufrshubham) [inline lib/game/enemy.dart]: Can be written as `position: size/2` for simplicity.

File: lib/game/player.dart [hunk 3]
```diff
@@ -68,25 +63,38 @@ class Player extends SpriteComponent
...
     super.onMount();
 
     // Adding a circular hitbox with radius as 0.8 times
     // the smallest dimension of this components size.
-    final shape = HitboxCircle(definition: 0.8);
-    addShape(shape);
+    final shape = CircleHitbox.relative(
+      0.8,
+      parentSize: this.size,
+      position: Vector2(size.x / 2, size.y / 2),
```
Comment (ufrshubham) [inline lib/game/player.dart]: Can be written as `position: size/2` for simplicity.

File: lib/game/power_ups.dart [hunk 4]
```diff
@@ -44,8 +44,13 @@ abstract class PowerUp extends SpriteComponent
   @override
   void onMount() {
     // Add a circular hit box for this power up.
-    final shape = HitboxCircle(definition: 0.5);
-    addShape(shape);
+    final shape = CircleHitbox.relative(
+      0.5,
+      parentSize: this.size,
+      position: Vector2(size.x / 2, size.y / 2),
```
Comment (ufrshubham) [inline lib/game/power_ups.dart]: Can be written as `position: size/2` for simplicity.

**Review by ufrshubham — CHANGES_REQUESTED** [review]
Everything looks good, but there seems to be some issue with pausing the game. If I pause the game, exit to main menu and then again start the game, the game loads up in the paused state of previous session. Can you see if you can fix that?

**Comment by rozpo** [comment]
> Everything looks good, but there seems to be some issue with pausing the game. If I pause the game, exit to main menu and then again start the game, the game loads up in the paused state of previous session. Can you see if you can fix that?

Fixed, now when player starts a new game game engine will be in running state. Moreover I handled another issue with collision detection between Player and Enemy. Collision was detected twice because callback set enemy hp = 0, and destroy was invoked only in next update cycle. In this situation Player and Enemy hit themselves twice.

**Review by ufrshubham — APPROVED** [review]
LGTM!

**Pull request merged**